File: src/app/contracts/responses.py
from typing import Any, Literal

from pydantic import BaseModel, Field


ConfidenceLevel = Literal["low", "medium", "high"]
RunStatus = Literal["queued", "running", "completed", "failed"]


class Evidence(BaseModel):
    source_type: str
    source_id: str | None = None
    locator: str
    snippet: str | None = None
    timestamp: str | None = None
    confidence: ConfidenceLevel = "medium"


class BaseWorkflowResponse(BaseModel):
    run_id: str
    status: RunStatus
    trace_id: str
    request_type: str
    primary_intent: str
    secondary_intents: list[str] = Field(default_factory=list)
    selected_agents: list[str] = Field(default_factory=list)
    model_version: str
    skill_versions: dict[str, str] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)
    confidence: ConfidenceLevel = "medium"


class PlanResponse(BaseWorkflowResponse):
    summary: str
    impacted_areas: list[str]
    implementation_plan: list[str]
    tests: list[str]
    risks: list[str]
    open_questions: list[str]
    evidence: list[Evidence]


class ReviewFinding(BaseModel):
    category: str
    severity: ConfidenceLevel
    message: str
    evidence: list[Evidence] = Field(default_factory=list)


class ReviewResponse(BaseWorkflowResponse):
    summary: str
    review_findings: list[ReviewFinding]
    missing_tests: list[str]
    risks: list[str]
    readiness_verdict: Literal["ready", "needs_changes", "blocked"]
    evidence: list[Evidence]


class TestPlanResponse(BaseWorkflowResponse):
    unit_tests: list[str]
    integration_tests: list[str]
    regression_targets: list[str]
    edge_cases: list[str]
    execution_order: list[str]


class RunDetailResponse(BaseModel):
    run_id: str
    status: str
    created_at: str
    completed_at: str | None = None
    request_type: str
    primary_intent: str
    secondary_intents: list[str] = Field(default_factory=list)
    selected_agents: list[str] = Field(default_factory=list)
    user_id: str
    repo_scope: list[str] = Field(default_factory=list)
    model_version: str
    skill_versions: dict[str, str] = Field(default_factory=dict)
    request: dict[str, Any]
    result: dict[str, Any]
    trace_id: str


class TraceStep(BaseModel):
    step_name: str
    step_order: int
    status: str
    started_at: str
    ended_at: str
    latency_ms: int
    tool_calls: int
    confidence: ConfidenceLevel | None = None
    input_ref: str | None = None
    output_ref: str | None = None
    error_message: str | None = None


class ToolCallRecord(BaseModel):
    step_name: str
    tool_name: str
    status: str
    started_at: str
    ended_at: str
    duration_ms: int
    input_summary: str
    output_count: int
    error_message: str | None = None


class TraceResponse(BaseModel):
    trace_id: str
    steps: list[TraceStep]
    spans: list[TraceStep]
    tool_calls: list[ToolCallRecord] = Field(default_factory=list)
    exported_at: str | None = None
    error_summary: str | None = None


class FeedbackResponse(BaseModel):
    feedback_id: str
    stored: bool

File: frontend/src/api/contracts/responses.ts
export type ConfidenceLevel = "low" | "medium" | "high";
export type RunStatus = "queued" | "running" | "completed" | "failed" | "canceled";
export type ConnectorHealthStatus = "ok" | "degraded" | "unconfigured";

export interface Evidence {
  source_type: string;
  source_id?: string | null;
  locator: string;
  snippet?: string | null;
  timestamp?: string | null;
  confidence: ConfidenceLevel;
}

export interface BaseWorkflowResponse {
  run_id: string;
  status: RunStatus;
  trace_id: string;
  request_type: string;
  primary_intent: string;
  secondary_intents: string[];
  selected_agents: string[];
  model_version: string;
  skill_versions: Record<string, string>;
  prompt_versions: Record<string, string>;
  warnings: string[];
  confidence: ConfidenceLevel;
}

export interface PlanResponse extends BaseWorkflowResponse {
  summary: string;
  impacted_areas: string[];
  implementation_plan: string[];
  tests: string[];
  risks: string[];
  open_questions: string[];
  evidence: Evidence[];
}

export interface ReviewFinding {
  category: string;
  severity: ConfidenceLevel;
  message: string;
  evidence: Evidence[];
}

export interface ReviewResponse extends BaseWorkflowResponse {
  summary: string;
  review_findings: ReviewFinding[];
  missing_tests: string[];
  risks: string[];
  readiness_verdict: "ready" | "needs_changes" | "blocked";
  evidence: Evidence[];
}

export interface TestPlanResponse extends BaseWorkflowResponse {
  unit_tests: string[];
  integration_tests: string[];
  regression_targets: string[];
  edge_cases: string[];
  execution_order: string[];
}

export interface RunDetailResponse {
  run_id: string;
  status: string;
  created_at: string;
  completed_at?: string | null;
  request_type: string;
  primary_intent: string;
  secondary_intents: string[];
  selected_agents: string[];
  user_id: string;
  repo_scope: string[];
  model_version: string;
  skill_versions: Record<string, string>;
  prompt_versions: Record<string, string>;
  request: Record<string, unknown>;
  result: Record<string, unknown>;
  trace_id: string;
}

export interface TraceStep {
  step_name: string;
  step_order: number;
  status: string;
  started_at: string;
  ended_at: string;
  latency_ms: number;
  tool_calls: number;
  confidence?: ConfidenceLevel | null;
  input_ref?: string | null;
  output_ref?: string | null;
  error_message?: string | null;
}

export interface ToolCallRecord {
  step_name: string;
  tool_name: string;
  status: string;
  started_at: string;
  ended_at: string;
  duration_ms: number;
  input_summary: string;
  output_count: number;
  error_message?: string | null;
}

export interface TraceResponse {
  trace_id: string;
  steps: TraceStep[];
  spans: TraceStep[];
  tool_calls: ToolCallRecord[];
  metadata: Record<string, unknown>;
  exported_at?: string | null;
  error_summary?: string | null;
}

export interface FeedbackResponse {
  feedback_id: string;
  stored: boolean;
}

export interface ConnectorHealth {
  status: ConnectorHealthStatus;
  detail?: string | null;
}

export interface HealthResponse {
  status: "ok" | "degraded";
  version: string;
  connectors: Record<string, ConnectorHealth>;
}

export interface ErrorResponse {
  code: string;
  message: string;
  request_id: string;
  detail?: unknown;
}
